diesel/src/sqlite/connection/raw.rs

```diff
@@ -292,99 +313,160 @@ extern "C" fn run_aggregator_step_function<ArgsSqlType, RetSqlType, Args, Ret, A
-            },
-            None => None,
-        };
-
-        let result = A::finalize(aggregator);
+        ffi::sqlite3_aggregate_context(ctx, 0)
+    };
 
-        match process_sql_function_result::<RetSqlType, Ret>(result) {
-            Ok(value) => value.result_of(ctx),
-            Err(e) => {
-                let msg = e.to_string();
-                ffi::sqlite3_result_error(ctx, msg.as_ptr() as *const _, msg.len() as _);
+    let mut aggregate_context = NonNull::new(aggregate_context as *mut OptionalAggregator<A>);
+    let aggregator = aggregate_context.as_mut().map(|a| {
+        let a = unsafe { a.as_mut() };
+        match std::mem::replace(a, OptionalAggregator::None) {
+            OptionalAggregator::Some(agg) => agg,
+            OptionalAggregator::None => {
+                eprintln!("We've written to the aggregator in the xStep callback. If xStep was never called, then ffi::sqlite_aggregate_context() would have returned a NULL pointer");
+                std::process::abort();
             }
         }
+    });
+
+    let result = std::panic::catch_unwind(|| A::finalize(aggregator))
+        .map(process_sql_function_result::<RetSqlType, Ret>);
+
+    match result {
+        Ok(Ok(value)) => value.result_of(ctx),
+        Ok(Err(e)) => {
+            let msg = e.to_string();
+            unsafe { context_error_str(ctx, &msg) };
+        }
+        Err(_) => {
+            let msg = format!("{}::finalize() panicked", std::any::type_name::<A>());
+            unsafe { context_error_str(ctx, &msg) };
+        }
     }
 }
 
 unsafe fn null_aggregate_context_error(ctx: *mut ffi::sqlite3_context) {
     static NULL_AG_CTX_ERR: &str = "An unknown error occurred. sqlite3_aggregate_context returned a null pointer. This should never happen.";
 
-    ffi::sqlite3_result_error(
-        ctx,
-        NULL_AG_CTX_ERR.as_ptr() as *const _ as *const _,
-        NULL_AG_CTX_ERR.len() as _,
-    );
+    context_error_str(ctx, NULL_AG_CTX_ERR)
+}
+
+unsafe fn context_error_str(ctx: *mut ffi::sqlite3_context, error: &str) {
+    ffi::sqlite3_result_error(ctx, error.as_ptr() as *const _, error.len() as _);
 }
 
-extern "C" fn destroy_boxed_fn<F>(data: *mut libc::c_void)
+#[allow(warnings)]
+extern "C" fn run_collation_function<F>(
+    user_ptr: *mut libc::c_void,
+    lhs_len: libc::c_int,
+    lhs_ptr: *const libc::c_void,
+    rhs_len: libc::c_int,
+    rhs_ptr: *const libc::c_void,
+) -> libc::c_int
 where
-    F: FnMut(&RawConnection, &[*mut ffi::sqlite3_value]) -> QueryResult<SerializedValue>
-        + Send
-        + 'static,
+    F: Fn(&str, &str) -> std::cmp::Ordering + Send + 'static,
 {
+    let user_ptr = user_ptr as *const F;
+    let f = unsafe { user_ptr.as_ref() }.unwrap_or_else(|| {
+        eprintln!(
+            "An unknown error occurred. user_ptr is a null pointer. This should never happen."
```

Comment:
Could you improve that error message saying something like: If you ever see this message, please open an issue at the diesel github repository?

(Maybe with an indication on which line this error happened?)